grpc_adapter.go

```diff
@@ -0,0 +1,668 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connect
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+)
+
+type grpcAdapterConfig struct {
+	Codec         Codec
```

Comment:
This is not needed. GRPC trailers _always_ use the proto format for encoding the status. So just use `proto.Marshal` and `proto.Unmarshal` instead of this, which (to my eyes) implies that the trailer statuc codec could be customized (which is not true with GRPC).

Reply:
It's part of the current internal func `grpcErrorFromTrailer`.

Reply:
If we're factoring this out into a separate package or repo, we don't need this flexibility. It's there in the main code path in case the user prefers a different implementation of protobuf serialization (vtproto, csproto, etc.).

In retrospect, using user-supplied codecs here may have been a bad idea - it complicates custom codecs a fair bit, since they need to handle both application-specific request/responses (where users have wide latitude to do unusual things) and protocol internals (where users have very little freedom to change behavior).

Reply:
Separated it into package `grpcadapter` with go import: github.com/connect-go/grpc-adapter.